lib/semantics/tools.cc

```diff
@@ -327,6 +335,28 @@ const Symbol *FindFunctionResult(const Symbol &symbol) {
   return nullptr;
 }
 
+// Return the variable of a construct association, if associated
+// Return nullptr if the name is associated with an expression
+const Symbol *GetAssociationRoot(const Symbol &symbol) {
+  const Symbol &ultimate{symbol.GetUltimate()};
+  if (ultimate.has<AssocEntityDetails>()) {  // construct association
+    auto &details{ultimate.get<AssocEntityDetails>()};
+    if (const MaybeExpr * pexpr{&details.expr()}) {
```

Comment:
If pexpr is false, `ultimate` will be returned. I think it should be nullptr in that case.

Reply:
Yes.  I'll restructure this to be correct and more obviously so.